tests/integration/rbac_test.go

```diff
@@ -0,0 +1,122 @@
+
+		// TODO: this needs to be changed when service-mesh api changes
+		it("rio-standard should not be able to create service-mesh services", func() {
+			testService.Kubeconfig = standardUser.GetKubeconfig()
+			err := testService.CreateWithError(t, "nginx")
+			if err == nil {
+				t.Fatal("rio-standard should not be able to create service that enable service mesh")
+			}
+			assert.True(t, strings.Contains(err.Error(), insuffienctPrivilegesMsg))
+			assert.True(t, strings.Contains(err.Error(), "rio-servicemesh"))
+		})
+
+		// TODO: we need to add more test for custom verb: hostport, hostnet, hostmount, serviceMesh and privilege
+		it("rio-standard should not be able to create host-network services", func() {
+			testService.Kubeconfig = standardUser.GetKubeconfig()
+			err := testService.CreateWithError(t, "--net", "host", "nginx")
+			if err == nil {
+				t.Fatal("rio-standard should not be able to create service that enable host networking")
+			}
+			assert.True(t, strings.Contains(err.Error(), insuffienctPrivilegesMsg))
+			assert.True(t, strings.Contains(err.Error(), "rio-hostnetwork"))
+		})
+
+		it("rio-readonly should not be able to create services", func() {
+			testService.Kubeconfig = readonlyUser.GetKubeconfig()
+			err := testService.CreateWithError(t, "nginx")
+			if err == nil {
+				t.Fatal("rio-readonly user should not be able to create services")
+			}
+			assert.True(t, strings.Contains(err.Error(), "is forbidden"))
+		})
+
+		it("rio-standard user should not be able to escalate privilege on global permissions", func() {
+			testService.Kubeconfig = standardUser.GetKubeconfig()
+			err := testService.CreateWithError(t, "--global-permission", "list rio.cattle.io/services", "nginx")
+			if err == nil {
+				t.Fatal("rio-standard should not be able to escalate privilege on global permissions")
+			}
+			assert.True(t, strings.Contains(err.Error(), insuffienctPrivilegesMsg))
+		})
+
+		it("rio-standard user should not be able to create privileges it doesn't have in the current namespace", func() {
+			testService.Kubeconfig = standardUser.GetKubeconfig()
+			err := testService.CreateWithError(t, "--permission", "update admin.rio.cattle.io/publicdomain", "--disable-service-mesh", "nginx")
+			if err == nil {
+				t.Fatal("rio-standard should not be able to create privileges it doesn't has")
+			}
+			assert.True(t, strings.Contains(err.Error(), insuffienctPrivilegesMsg))
+		})
+
+		it("rio-standard user should be able to create privileges it already has 1", func() {
+			testService.Kubeconfig = standardUser.GetKubeconfig()
+			testService.Create(t, "--permission", "list rio.cattle.io/services", "--disable-service-mesh", "nginx")
+		})
+
+		it("rio-standard user should be able to create privileges it already has 2", func() {
+			testService.Kubeconfig = standardUser.GetKubeconfig()
+			testService.Create(t, "--permission", "watch rio.cattle.io/app", "--disable-service-mesh", "nginx")
+		})
+
+	}, spec.Parallel(), spec.Flat())
```

Comment:
Curious if there is a reason you're using spec.flat ? Thats the default.

Reply:
Noticed that without this Flat the parent context(Before) will be executed on every run.

Reply:
I believe the before block will run on every `it` block with either flat or nested. Afaik flat just means that all these tests aren't sub-scoped and can be run in parallel with any other suite's test.